Issue/282/modeling structure

vitenti opened the pull request:
This PR solves several issues:

- #144 cluster_toolkit is one of the possible backends, I don't think it should be included as a submodule/subproject since now we support other backends (CCL and NumCosmo) that are in active development. If all agree this should close that issue.
- #207 this issue concerns the different possible background mass definition (mean matter, critical, etc), this has out-of-the-box support in both CCL and NumCosmo. Therefore, this PR fixes this issue.
- #254 The comparison between NumCosmo and CLMM no longer makes sense since NumCosmo is now one of the possible backends. The comparison was in fact between cluster_toolkit and NumCosmo, there is a comparison notebook (CCL/Colossus/NumCosmo) in NumCosmo repo ( https://github.com/NumCosmo/NumCosmo/blob/master/notebooks/NC_CCL_Colossus_density_profiles.ipynb ).
- #267 Fixed by this PR.
- #277 Also fixed by this PR see #254 above.
- #282 The issue that originated this PR. The modeling module now is a script that loads from different backends for the functional interface that was developed in CLMM. It works transparently and load the backend depending on the value of a environment variable CLMM_MODELING_BACKEND (ct => cluster_toolkit, nc => numcosmo, ccl => ccl) . A OO interface was also included for CCL and NumCosmo. All modeling unit tests are now applied to all available backends. The actual backends are in the modbackend directory.
- #287 fixed by this PR.
- #296 also fixed here (also makes branch issue/254/clmm_numcosmo_cmp obsolete).

vitenti commented:
This PR requires NumCosmo version >= 0.15.2 to use its backend. (It is already available in conda-forge).

vitenti commented:
The drop on the coverage (coveralls) results from some backends (NumCosmo and CCL) not being available in the Coveralls build.

combet commented:
Thanks for the thorough PR @vitenti. Two quick questions/comments:
- I did not see any changes to the API documentation and noticed you exlcuded `api/clmm.modbackend.*` in `docs/conf.py`. Is that because these files are not yet documented?
- Regarding the drop in coverage, I understand it's because we're only installing cluster-toolkit in `.travis.yml` and not Numcosmo and CCL as well? Could we try adding a conda install for NC and CCL as well? I fear this will make the CI process a bit heavy though. What do you think?

combet commented:
Also, here're a few things/questions/comments, in no particular order (I'll update the list as I progress through the code)
-  in the OO approach, it would be useful to add a `backend` attribute to the `Modeling` object so that the user may easily check what backend they're using. e.g.
   - instantiating `moo = clmm.Modeling (massdef = 'mean', delta_mdef = 200, halo_profile_model = 'nfw')`,
   - `moo.backend` would then return 'nc' or 'ccl' depending on what has actually been used.

vitenti commented:
Hello @combet , I implemented your idea of having a property showing which backend is being used. Moreover, I just included CCL and NumCosmo to the travis build, it now takes about 5min, and still is a reasonable building time in my opinion. That fixed the coverage problem. 

About the documentation, we still need to decide the best way to document things (for instance, just one documentation in the abstract interface or one for each backend). That's why I removed these files from the doc building for the moment.

combet commented:
Thanks @vitenti for the latest modifications! Great to see Travis working with the 3 backends and coverage back up!

> About the documentation, we still need to decide the best way to document things (for instance, just one documentation in the abstract interface or one for each backend). That's why I removed these files from the doc building for the moment.

Thanks for clarifying. So I think this should be another issue to open once this PR is merged.

combet commented:
I also noticed the following in the OO modeling notebook:
- [ ] You have `DeltaSigma = moo.eval_sigma_mean (r3d, z_cl)`, but it should be DeltaSigma = moo.eval_sigma_excess (r3d, z_cl), right?

combet commented:
@vitenti, I ran into an issue when generating mock data with the `ccl` background in the "object oriented "`Example2..._oo.ipynb` (cell 7) that I added to this branch. I copy the end of the error message below. This occurs when asking for "photoz errors" on the mock data, i.e., with the line 

`noisy_data_z = mock.generate_galaxy_catalog(cluster_m, cluster_z, concentration, cosmo, ngals, Delta,'chang13', 
                                           shapenoise=0.05, 
                                           photoz_sigma_unscaled=0.05)`

![Screenshot 2020-09-14 at 14 33 24](https://user-images.githubusercontent.com/20227809/93086324-40731580-f697-11ea-974e-96e194f9a40e.png)

This runs fine with the `nc` backend. I will try to troubleshoot this further before this end of the week. Posting it here for reference, and any insights are very welcome !

vitenti commented:
Hello @combet it seems that the problem comes from the way things are vectorized in CCL, we must check if it is consistent with the way we are expecting things to be vectorized in CLMM. That's something that we should standardize.

vitenti commented:
For those interested I included two examples using NumCosmo statistical framework, you can find them in the branch issue/282/numcosmo_example the notebook Example2_Fit_Halo_Mass_to_Shear_Catalog_OO_NumCosmo.ipynb has a full MCMC analysis using both the wrong and correct approach. There I implemented the non-binned likelihood, I still didn't have time to implement the correct projected likelihood, but apart from the bias we can have an idea on how different will be the constraints using the two methods.

m-aguena commented on clmm/modbackend/ccl.py:
```diff
@@ -0,0 +1,164 @@
+# Functions to model halo profiles 
```
Can you add docstrings to the fucntions here? If the API here work exactly as in `clmm/modbackend/func_layer.py`, you could just copy them.

vitenti replied:
In this file we have the OO interface to CCL, thus, it will inherit the docstring from CLMModeling, for this reason it is better to put all documentation in clmm_modeling.py and only add a docstring to the methods in ccl.py when we want to overwrite the father's documentation.

For example, lets say that in clmm_modeling.py we add a docstring documenting the behavior of set_mass, since CCLCLMModeling is a child of CLMModeling its implementation of set_mass will inherit the docstring from CLMModeling's set_mass method. If you want to see this working just use:

```
import inspect
print(inspect.getdoc(moo.set_mass))
```

and you will see the docstring that I just added to set_mass in clmm_modeling.py.

Since the documentation in func_layer.py concerns the functional interface they will be a bit different from the OO documentation. But here we have the same as in the oo interface, we just need to document the func_layer once and all back-ends will inherit it.

m-aguena commented on clmm/modbackend/numcosmo.py:
```diff
@@ -0,0 +1,186 @@
+# NumCosmo implementation of CLMModeling
```
Can you add docstrings to the fucntions here? If the API here work exactly as in `clmm/modbackend/func_layer.py`, you could just copy them.

vitenti replied:
See above

m-aguena commented on clmm/modbackend/func_layer.py:
```diff
@@ -0,0 +1,509 @@
+# Thin functonal layer on top of the class implementation of CLMModeling .
```
Is there a reason for this and `generic.py` being separated files?

vitenti replied:
Yes, func_layer.py  is just a thin layer providing a functional interface for the OO CLMModeling object (implemented by an actual back-end). The file generic.py contains all functions that can be used by any back-end.

m-aguena commented on clmm/modbackend/ccl.py:
```diff
@@ -0,0 +1,164 @@
+import warnings
+
+from . import func_layer
+from . func_layer import *
```
please do not import `*`. Either import a module and use it or make specific imports

vitenti replied:
In general that's a good advice but in this case we are using the __all__ variable to control the exports from func_layer. That way, we do not need to repeat the functions names in all back-ends, and once a new function in added to func_layer __all__ variable, all back-ends will have it automatically.

m-aguena commented on clmm/modbackend/numcosmo.py:
```diff
@@ -0,0 +1,186 @@
+import warnings
+
+from . import func_layer
+from . func_layer import *
```
please do not import `*`. Either import a module and use it or make specific imports

vitenti replied:
See above.

m-aguena commented on clmm/modbackend/cluster_toolkit.py:
```diff
@@ -0,0 +1,562 @@
+from ..constants import Constants as const
+from .cluster_toolkit_patches import _patch_comoving_coord_cluster_toolkit_rho_m
+from . import generic
+from . generic import *
```
please do not import `*`. Either import a module and use it or make specific imports

vitenti replied:
See above.

m-aguena commented on clmm/modbackend/clmm_modeling.py:
```diff
@@ -0,0 +1,36 @@
+# CLMModeling abstract class
```
If this parent class does not has any pre-set function, is it worth it to keep it? It just serve as an API template as it is.

vitenti replied:
Using an abstract base class is good to control the subsequent implementations, take a look at [PEP 3119](https://www.python.org/dev/peps/pep-3119). Here we can put all documentation and common implementation of methods. We should actually make it better by using ABC in clmm_modeling.py .

aimalz replied:
I think another way to phrase this is that we need it to support the back-ends just because they have modeling objects that behave differently, similarly to why we need a cosmology object (#293).

combet reviewed (approved):
It looks like @m-aguena will most likely be unavailable this week. @vitenti replied to all the points. Approving this PR. Will merge later in the week.

combet commented:
@vitenti  -  Regarding the docstrings, I saw the example you provided in `clmm_modeling` for `set_mass`, but the other functions are not documented yet. Do you agree we can open a separate issue and address this once this PR is merged? (We really need to get this PR through so we can move to other things that depend on it.)

vitenti commented:
@combet Yes, absolutely, we should merge ASAP.
